apps/ec.c

```diff
@@ -188,105 +177,107 @@ int ec_main(int argc, char **argv)
     }
 
     BIO_printf(bio_err, "read EC key\n");
-    if (informat == FORMAT_ASN1) {
-        if (pubin)
-            eckey = d2i_EC_PUBKEY_bio(in, NULL);
-        else
-            eckey = d2i_ECPrivateKey_bio(in, NULL);
-    } else if (informat == FORMAT_ENGINE) {
-        EVP_PKEY *pkey;
-        if (pubin)
-            pkey = load_pubkey(infile, informat, 1, passin, e, "public key");
-        else
-            pkey = load_key(infile, informat, 1, passin, e, "private key");
-        if (pkey != NULL) {
-            eckey = EVP_PKEY_get1_EC_KEY(pkey);
-            EVP_PKEY_free(pkey);
-        }
-    } else {
-        if (pubin)
-            eckey = PEM_read_bio_EC_PUBKEY(in, NULL, NULL, NULL);
-        else
-            eckey = PEM_read_bio_ECPrivateKey(in, NULL, NULL, passin);
-    }
+
+    if (pubin)
+        eckey = load_pubkey(infile, informat, 1, passin, e, "public key");
+    else
+        eckey = load_key(infile, informat, 1, passin, e, "private key");
+
     if (eckey == NULL) {
         BIO_printf(bio_err, "unable to load Key\n");
-        ERR_print_errors(bio_err);
         goto end;
     }
 
     out = bio_open_owner(outfile, outformat, private);
     if (out == NULL)
         goto end;
 
-    group = EC_KEY_get0_group(eckey);
-
-    if (new_form)
-        EC_KEY_set_conv_form(eckey, form);
+    if (point_format
+        && !EVP_PKEY_set_utf8_string_param(
+                eckey, OSSL_PKEY_PARAM_EC_POINT_CONVERSION_FORMAT,
+                point_format)) {
```

Comment:
With the #13913 it will. So perhaps add a TODO for now? Or should we wait with the merge until the #13913 is merged?

It does not make much sense to me to create a set of new helper functions for kind-of corner-case uses that won't be truly needed later.

Reply:
> With the #13913 it will.

Errrrrr.....  that's not a given.  I just went looking to see how this was done in legacy, and the answer is "not at all", at least in form of ctrl calls.  There was only one way, to have direct access to the `EV_KEY` and calling `EC_KEY_set_conv_form(eckey, form)` (which is what pre-3.0 OpenSSL does).

To support this for legacy keys, we would need to start with adding support for it via `EVP_PKEY_CTX_ctrl()` (or possibly `EVP_PKEY_ctrl()`, which is more work as that would be completely new functionality), i.e. modify the ctrl functions in `crypto/ec/ec_ameth.c` or `crypto/ec/ec_pmeth.c`.  When *that* in place, a translation can be added in #13913.

EC_KEY support has been found wanting at an EVP level for a loooong time, and this is one of the cases where that shows.

Reply:
Hmmm....I see little point in *adding* new ctrls for this. Does that mean that if we have an ENGINE that uses an EC_KEY we can't support it?

Reply:
> Does that mean that if we have an ENGINE that uses an EC_KEY we can't support it for this functionality?

That's exactly what it means.

Reply:
Ok. I guess we have to live with that.

Reply:
We might want to document this...  it does, after all, remove an existing functionality when engines are involved...
(mind you, EC_KEY support has always been lacking in ENGINEs...  for example, this has made it impossible to support EC_KEYs hidden behind an HSM)

Reply:
Added a line to HISTORY section of the ec(1) manpage.